src/main/kotlin/no/nav/faktureringskomponenten/service/beregning/AntallBeregner.kt

```diff
@@ -0,0 +1,64 @@
+package no.nav.faktureringskomponenten.service.beregning
+
+import mu.KotlinLogging
+import java.math.BigDecimal
+import java.math.RoundingMode
+import java.time.LocalDate
+import java.time.temporal.ChronoUnit
+
+private val log = KotlinLogging.logger { }
+
+class AntallBeregner {
+    companion object {
+        fun antallForPeriode(fom: LocalDate, tom: LocalDate): BigDecimal {
```

Comment:
Se min andre kommentar først. Jeg tror det er feil. Men hvis det ender med at din metode er mer riktig, så tror jeg du kan gjøre din kode mye enklere:
```
fun beløpForPeriode(enhetsprisPerMåned: BigDecimal, fraOgMed: LocalDate, tilOgMed: LocalDate): BigDecimal {
    val periode = Period.between(fraOgMed, tilOgMed)
    val antallDagerIPeriodensMåneder = ChronoUnit.DAYS.between(fraOgMed.withDayOfMonth(1), tilOgMed.with(TemporalAdjusters.lastDayOfMonth()))
    val antallDagerIPeriode = ChronoUnit.DAYS.between(fraOgMed, tilOgMed)
    val antallMåneder = periode.months + 1
    
    val totalBeløpForAlleMåneder = enhetsprisPerMåned.multiply(BigDecimal(antallMåneder))
    
    val result = totalBeløpForAlleMåneder.multiply(
        antallDagerIPeriode.toBigDecimal().divide(
            antallDagerIPeriodensMåneder.toBigDecimal(),
            2, RoundingMode.HALF_UP
        )
    )
    
    return result
}
```

Reply:
Hvis derimot min beregning skulle stemme, så burde dette funke:
```
fun beløpForPeriode(enhetsprisPerMåned: BigDecimal, datoFra: LocalDate, datoTil: LocalDate): BigDecimal {
    var totalpris = BigDecimal.ZERO
    var gjeldendeDato = datoFra
    while (gjeldendeDato <= datoTil) {
        val dagerIMåneden = gjeldendeDato.lengthOfMonth()
        val prisProsent: BigDecimal

        if (gjeldendeDato.month == datoFra.month && gjeldendeDato.month == datoTil.month) {
            val dagerIPerioden = datoTil.dayOfMonth - datoFra.dayOfMonth + 1
            prisProsent = BigDecimal(dagerIPerioden.toDouble() / dagerIMåneden.toDouble())
        } else if (gjeldendeDato.month == datoFra.month) {
            val dagerIPerioden = dagerIMåneden - datoFra.dayOfMonth + 1
            prisProsent = BigDecimal(dagerIPerioden.toDouble() / dagerIMåneden.toDouble())
        } else if (gjeldendeDato.month == datoTil.month) {
            prisProsent = BigDecimal(datoTil.dayOfMonth.toDouble() / dagerIMåneden.toDouble())
        } else {
            prisProsent = BigDecimal.ONE
        }

        totalpris += enhetsprisPerMåned * prisProsent
        gjeldendeDato = gjeldendeDato.plusMonths(1).withDayOfMonth(1)
    }

    return totalpris
}
```

Reply:
sorry, ble veldig nysjerrig på denne problemstillingen, så sjekket bare for å se om jeg kunne gjøre det enklere. :laughing:  Tror man kan gjøre det slik også, uten loops og mye ifs:

```
fun belopForPeriode(
    enhetsprisPerMåned: BigDecimal,
    fraOgMed: LocalDate,
    tilOgMed: LocalDate
): BigDecimal {
    // Hvis perioden er innenfor samme måned og år
    if (fraOgMed.month == tilOgMed.month && fraOgMed.year == tilOgMed.year) {
        val dagerMellom = BigDecimal(ChronoUnit.DAYS.between(fraOgMed, tilOgMed))
        val lengdePåMåned = BigDecimal(fraOgMed.lengthOfMonth())
        return dagerMellom.divide(lengdePåMåned, 10, RoundingMode.HALF_UP) * enhetsprisPerMåned
    } else {
        // Beregn prisen for første måned
        val sisteDagIFørsteMåned = fraOgMed.with(TemporalAdjusters.lastDayOfMonth())
        val dagerIFørsteMåned = BigDecimal(ChronoUnit.DAYS.between(fraOgMed, sisteDagIFørsteMåned) + 1)
        val prisForFørsteMåned = dagerIFørsteMåned.divide(BigDecimal(fraOgMed.lengthOfMonth()), 10, RoundingMode.HALF_UP) * enhetsprisPerMåned

        // Beregn prisen for siste måned
        val førsteDagISisteMåned = tilOgMed.withDayOfMonth(1)
        val dagerISisteMåned = BigDecimal(ChronoUnit.DAYS.between(førsteDagISisteMåned, tilOgMed) + 1)
        val prisForSisteMåned = dagerISisteMåned.divide(BigDecimal(tilOgMed.lengthOfMonth()), 10, RoundingMode.HALF_UP) * enhetsprisPerMåned

        // Beregn antall måneder mellom start- og sluttdatoene
        val totaltAntallMåneder = BigDecimal(Period.between(fraOgMed, tilOgMed).months + 1)

        // Beregn total kostnad
        val kostnad = (enhetsprisPerMåned * (totaltAntallMåneder - BigDecimal(2))) + prisForFørsteMåned + prisForSisteMåned

        return kostnad
    }
}

```

Reply:
100 måter man kan gjøre dette på, ved et raskt øyekast synes jeg dette er mer eller mindre leselig enn det vi allerede har. Men så er forslagene til beløp, og denne koden beregner antall for perioden (antallForPeriode) :smile_cat: Vi kan ikke ha det i samme kode uansett fordi de (OEBS) _skal_ ha antall og enhetspris, og trenger ikke beløp (men vi ønsker å beholde beløp for vår egen del), og det gjør at vi kan oppdage feil raskere hvis vi plutselig regner 100kr, og de regner 101kr

Reply:
På forhånd: Det kan være litt vanskelig å komme til en enighet her; Du skrev den andre koden, og for deg er det klarere. Samtidig er det jeg skrev mye klarere for meg fordi jeg skrev det. Obviously. Men jeg skriver bare hva jeg mener likevel, så du får avgjøre om du ønsker å ta med eller ikke - jeg er ikke i denne fokusgruppen og kan ikke bruke for mye tid.

> ved et raskt øyekast synes jeg dette verken er mer eller mindre leselig enn det vi allerede har

Jeg er uenig i det. AntallBeregner sammenliknet med siste innslaget mener jeg det er stor forskjell i lesbarhet, selv om funksjonaliteten i praksis er akkurat det samme. Jeg føler at når man tar med loops, så blir ting automatisk mer komplisert i hodet. Hele grunnen til at du har loops er å beregne antall HELE måneder i perioden det er. Ellers er antall dager i første og siste måned alltid lett å beregne uten loops. Antall måneder i mellom er så å si en "oneliner":
```
val månederMellom = if (erSammeMånedOgÅr) {
        BigDecimal.ZERO
    } else {
        ChronoUnit.MONTHS.between(datoFra.plusMonths(1).withDayOfMonth(1), datoTil).toBigDecimal()
    }
```
 Her er det enkel og straight forward vei til samme svar, uten å ha mange variabler i hodet samtidig.

Men igjen, du får selv avgjøre.

Reply:
> trenger ikke beløp (men vi ønsker å beholde beløp for vår egen del), og det gjør at vi kan oppdage feil raskere hvis vi plutselig regner 100kr, og de regner 101kr

Koden beregner beløp slik:
```
    val totaltProsent = førsteMånedProsent + månederMellom + sisteMånedProsent
    return enhetsprisPerMåned * totaltProsent
```

Dropp siste linja og returner totaltProsent, så har du prosentene.

Reply:
Ja det var ikke en dårlig idé! Skal jobbe litt med repo+domene problemet i dag, men kanskje vi kan ta litt parprog på dette i morgen hvis jeg ikke blir tidlig ferdig. :star:

Reply:
:heavy_check_mark:  Brukte det siste forslaget ditt som basis for forbedringen! Nå er koden mye lettere å forstå og kan endres ved behov ! :star: :star: